sdcflows/transform.py

```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        allow_negative : :obj:`bool`
+            Remove negative values introduced in interpolation (may happen for nonnegative data
+            when order :math:`\gt` 3). Set this value to `True` if your `moving` image does
+            have negative values.
 
         Returns
         -------
-        resampled : `spatialimage` or ndarray
+        resampled : :obj:`~nibabel.spatialimages.SpatialImage`
             The data imaged after resampling to reference space.
 
         """
-        from sdcflows.utils.tools import ensure_positive_cosines
 
         # Ensure the fmap has been computed
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
+        if isinstance(moving, (str, bytes, Path)):
+            moving = nb.load(moving)
 
-        spatialimage, axcodes = ensure_positive_cosines(spatialimage)
+        # Make sure the data array has all cosines positive (i.e., no axes are flipped)
+        moving, axcodes = ensure_positive_cosines(moving)
 
-        self.fit(spatialimage)
-        fmap = self.mapped.get_fdata().copy()
-
-        # Reverse mapped if reversed blips
-        if pe_dir.endswith("-"):
-            fmap *= -1.0
-
-        # Generate warp field
-        pe_axis = "ijk".index(pe_dir[0])
-
-        # Map voxel coordinates applying the VSM
-        if self.xfm is None:
-            ijk_axis = tuple([np.arange(s) for s in fmap.shape])
-            voxcoords = np.array(
-                np.meshgrid(*ijk_axis, indexing="ij"), dtype="float32"
-            ).reshape(3, -1)
+        if self.mapped is not None:
+            warn(
+                "The fieldmap has been already fit, the user is responsible for "
+                "ensuring the parameters of the EPI target are consistent."
+            )
         else:
-            # Map coordinates from reference to time-step
-            self.xfm.reference = spatialimage
-            hmc_xyz = self.xfm.map(self.xfm.reference.ndcoords.T)
-            # Convert from RAS to voxel coordinates
-            voxcoords = (
-                np.linalg.inv(self.xfm.reference.affine)
-                @ _as_homogeneous(np.vstack(hmc_xyz), dim=self.xfm.reference.ndim).T
-            )[:3, ...]
-
-        # fmap * ro_time is the voxel-shift map (VSM)
-        # The VSM is just the displacements field given in index coordinates
-        # voxcoords is the deformation field, i.e., the target position of each voxel
-        voxcoords[pe_axis, ...] += fmap.reshape(-1) * ro_time
-
-        # Prepare data
-        data = np.squeeze(np.asanyarray(spatialimage.dataobj))
-        output_dtype = output_dtype or data.dtype
+            # Generate warp field (before ensuring positive cosines)
+            self.fit(moving, xfm_data2fmap=xfm_data2fmap, approx=approx)
+
+        # Squeeze n33on-spatial dimensions
+        newshape = moving.shape[:3] + tuple(dim for dim in moving.shape[3:] if dim > 1)
+        data = np.reshape(moving.dataobj, newshape)
```

Comment:
If it saves anything to avoid loading data into RAM this early, you can reshape array proxies without loading them:

```suggestion
        data = nb.arrayproxy.reshape_dataobj(moving.dataobj, newshape)
```

Feel free to close if it doesn't make a difference.